src/ports/postgres/modules/graph/sssp.py_in

```diff
@@ -0,0 +1,347 @@
+					{vertex_id} INT, {weight} INT, parent INT) {distribution};
+			""",  <!''!>)
+		plpy.execute(sql.format(**locals()))
+
+		plpy.execute(
+			""" INSERT INTO {message} VALUES({source_vertex},0,{source_vertex})
+			""".format(**locals()))
+
+		v_cnt = plpy.execute("SELECT count(*) FROM {vertex_table}".
+			format(**locals()))[0]['count']
+		for i in range(0,v_cnt+1):
+
+			plpy.execute("TRUNCATE TABLE {0}".format(toupdate))
+			plpy.execute(
+				""" INSERT INTO {toupdate}
+					SELECT DISTINCT ON (message.id) message.id AS id,
+						message.val AS val,
+						message.parent AS parent
+					FROM {message} AS message, {out_table} AS out_table
+					WHERE message.id = out_table.{vertex_id}
+						AND message.val<out_table.{weight}
+				""".format(**locals()))
+
+			if table_is_empty(toupdate):
+				break
+			sql = m4_ifdef(<!__HAWQ__!>,
+				<!"""
+				TRUNCATE TABLE {temp_table};
+				INSERT INTO {temp_table}
+					SELECT *
+					FROM {out_table}
+					WHERE {out_table}.{vertex_id} NOT IN (
+						SELECT {toupdate}.id FROM {toupdate})
+					UNION
+					SELECT * FROM {toupdate};
+				DROP TABLE {out_table};
+				ALTER TABLE {temp_table} RENAME TO {out_table};
+				CREATE TABLE {temp_table} (
+					{vertex_id} INT, {weight} INT, parent INT) {distribution};
+				"""!>,
+				<!"""
+				UPDATE {out_table} SET
+				{weight}=toupdate.val,
+				parent=toupdate.parent
+				FROM
+				{toupdate} AS toupdate
+				WHERE
+				{out_table}.{vertex_id}=toupdate.id
+				"""!>)
+			plpy.execute(sql.format(**locals()))
+			plpy.execute("TRUNCATE TABLE {0}".format(message))
+			plpy.execute(
+				""" INSERT INTO {message}
+					SELECT edge_table.{dest} AS id, x.val AS val,
+						toupdate.id AS parent
+					FROM {toupdate} AS toupdate, {edge_table} AS edge_table, (
+						SELECT edge_table.{dest} AS id,
+							min(toupdate.val + edge_table.{weight}) AS val
+						FROM {toupdate} AS toupdate,
+							{edge_table} AS edge_table
+						WHERE edge_table.{src}=toupdate.id
+						GROUP BY edge_table.{dest}) x
+					WHERE edge_table.{src} = toupdate.id
+						AND edge_table.{dest} = x.id
+						AND toupdate.val + edge_table.{weight} = x.val
+						""".format(**locals()))
+
+		if i == v_cnt:
+			plpy.error("Graph SSSP: Detected a negative cycle in the graph.")
+
+		sql = m4_ifdef(<!__HAWQ__!>,
+			""" DROP TABLE {temp_table} """,  <!''!>)
+		plpy.execute(sql.format(**locals()))
+
+	return None
+
+def graph_sssp_get_path(schema_madlib, sssp_table, dest_vertex, **kwargs):
+	"""
+	Helper function that can be used to get the shortest path for a vertex
+    Args:
+    	@param source_table	Name of the table that contains the SSSP output.
+        @param out_table	The vertex that will be the destination of the
+            				desired path.
+	"""
+	validate_get_path(sssp_table, dest_vertex)
+	cur = dest_vertex
+	cols = get_cols(sssp_table)
+	id = cols[0]
+	ret = [dest_vertex]
+	sql = "SELECT parent FROM {sssp_table} WHERE {id} = {cur} LIMIT 1"
```

Comment:
This will almost certainly be faster if done in a recursive CTE. If for some reason that won't work, at least plan the query and re-use the plan.